src/groth16/mod.rs

```diff
@@ -466,6 +771,170 @@ impl<'a, E: Engine> ParameterSource<E> for &'a Parameters<E> {
     }
 }
 
+// A re-usable method for parameter loading via mmap.
+fn read_g1<E: Engine>(
+    mmap: &Mmap,
+    start: usize,
+    end: usize,
+    checked: bool,
+) -> Result<E::G1Affine, std::io::Error> {
+    let ptr = &mmap[start..end];
+    let repr =
+        unsafe { *(ptr as *const [u8] as *const <E::G1Affine as CurveAffine>::Uncompressed) };
+
+    if checked {
+        repr.into_affine()
+    } else {
+        repr.into_affine_unchecked()
+    }
+    .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+    .and_then(|e| {
+        if e.is_zero() {
+            Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "point at infinity",
+            ))
+        } else {
+            Ok(e)
+        }
+    })
+}
+
+// A re-usable method for parameter loading via mmap.
+fn read_g2<E: Engine>(
+    mmap: &Mmap,
+    start: usize,
+    end: usize,
+    checked: bool,
+) -> Result<E::G2Affine, std::io::Error> {
+    let ptr = &mmap[start..end];
+    let repr =
+        unsafe { *(ptr as *const [u8] as *const <E::G2Affine as CurveAffine>::Uncompressed) };
+
+    if checked {
+        repr.into_affine()
+    } else {
+        repr.into_affine_unchecked()
+    }
+    .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+    .and_then(|e| {
+        if e.is_zero() {
+            Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "point at infinity",
+            ))
+        } else {
+            Ok(e)
+        }
+    })
+}
+
+impl<'a, E: Engine> ParameterSource<E> for &'a MappedParameters<E> {
+    type G1Builder = (Arc<Vec<E::G1Affine>>, usize);
+    type G2Builder = (Arc<Vec<E::G2Affine>>, usize);
+
+    fn get_vk(&mut self, _: usize) -> Result<VerifyingKey<E>, SynthesisError> {
+        Ok(self.vk.clone())
+    }
+
+    fn get_h(&mut self, _num_h: usize) -> Result<Self::G1Builder, SynthesisError> {
+        let params = File::open(self.param_file.clone())?;
```

Comment:
shouldn't the mmaped params have a single file in the struct, instead of opening it new everytime these methods are called?

Reply:
Keeping around a persistent mmap'd file accessed via multiple threads is potentially dangerous and then we have to deal with possible re-mapping issues, etc.  I thought there was a comment explaining that a single map per accessed var (at this API level, such as 'get_h'), but perhaps not.  It appeared to still be a win in terms of run-time performance given the RAM reduction.